_source/_posts/2022-08-24-state-of-project-loom.md

```diff
@@ -0,0 +1,261 @@
+
+        scope.join();           // Join both forks
+        scope.throwIfFailed();  // ... and propagate errors
+
+        // Here, both forks have succeeded, so compose their results
+        System.out.println("Inventory " + inventory.resultNow() + " updated for order " + order.resultNow());
+    }
+}
+```
+
+Unlike the previous sample using `ExecutorService`, we can now use `StructuredTaskScope` to achieve the same result while confining the lifetimes of the subtasks to the lexical scope, in this case, the body of the `try-with-resources` statement. The code is much more readable, and the intent is also clear. `StructuredTaskScope` also ensures the below behavior automatically
+
+- **Error handling with short-circuiting** — If either the `updateInventory()` or `updateOrder()` fails, the other is cancelled unless its already completed. This is managed by the cancellation policy implemented by `ShutdownOnFailure()`; other policies are possible.
+
+- **Cancellation propagation** — If the thread running `handleOrder()` is interrupted before or during the call to `join()`, both forks are canceled automatically when the thread exits the scope.
+
+- **Observability** — A thread dump would clearly display the task hierarchy, with the threads running `updateInventory()` and `updateOrder()` shown as children of the scope.
+
+## State of Project Loom
+
+The project started in 2017 and has undergone many changes and proposals. Virtual threads were initially called fibers, but later on, it was renamed to avoid confusion. Today with Java 19 getting closer to release, the project has delivered two features discussed above. One as a preview and another as an incubator. Hence the path to stabilization of the features should be more precise.
+
+## What does this mean to average Java developers?
+
+When these features are production ready, it should not affect average Java developers much, as most java developers might be using libraries for most of the concurrency use cases. But it can be a big deal in those rare scenarios where you are doing a lot of multi-threading or parallel programming without using libraries. Virtual threads will increase performance and scalability in most cases, and structured concurrency can help simplify the codebase and make it less fragile and maintainable.
+
+## What does this mean to Java library developers?
+
+When these features are production ready, it will be a big deal for libraries and frameworks that use threads or parallelism. Library authors will see huge performance and scalability improvements while simplifying the codebase and making it maintainable. Web frameworks like Spring and Micronaut can switch to virtual threads to improve throughput further. I would expect most web frameworks and libraries to migrate to virtual threads from thread pools and the trendy reactive programming libraries like RxJava and Akka might be able to use structured concurrency effectively.
```

Comment:
What about foundational technologies like Tomcat? 

You might want to mention there will still be bottlenecks with blocking libraries like JDBC drivers. There will still be a need for reactive technologies that are non-blocking if you want maximum throughput.

Reply:
Ya true, but I would also expect the JDBC libs and frameworks also to adapt virtual threads were applicable but yes there will be use cases for reactive out there